### PR TITLE
i#2921: support asynch signal in late exit and when interrupting a clone syscall

### LINKED ISSUE #2921: code_api|pthreads.ptsig failing on x86 Travis
Seen once in https://github.com/DynamoRIO/dynamorio/pull/2906

```
 93/294 Test  #88: code_api|pthreads.ptsig ..........................................***Failed  Required regular expression not found.Regex=[^Estimation of pi is 3\.142425985001098
$
]  0.64 sec
<ERROR: master_signal_handler with no siginfo (i#26?): tid=19519, sig=10>
Estimation of pi is 3.142425985001098
```

### PR DESCRIPTION
**Pull request by hgreving2304**
Fixes an asynch signal arriving late when thread is on its way to exit by blocking all signals during exit. This is ok if because we're at the app's thread_exit system call. When doing a detach, the app's signal mask is restored before going native. For terminate events using kill, we are not blocking the kill signal. Suspend signal is also excluded because a detaching thread may try to synchronize with an exiting thread and as long as the signal is getting delivered to this thread, we need to reply to it from the signal handler.

Adds support to the signal handler for an asynch signal arriving in the middle of a clone system call or temporarily-native thread, while the spawning thread's tls magic sentinel is invalid. We're now detecting this condition in the signal handler by making assumptions as stated in this patch.

Fixes #2921

### CODE REVIEW

**Comment by derekbruening** [comment]
> This is ok if because we're past the app's thread_exit system call.

No, if we were past it the thread would not exist.  We are just prior to it.

**Comment by derekbruening** [comment]
> we need to be careful in the future that any thread has not been added to the global all_threads list as long as its tls magic is invalid.

I'm not sure I follow this: it sounds like you're adding an invariant that a thread on the all_threads list cannot have an invalid magic field, which is violated by design today.  Do you mean something else?  Maybe this needs to be re-worded.

**Comment by hgreving2304** [comment]
> > This is ok if because we're past the app's thread_exit system call.
> 
> No, if we were past it the thread would not exist. We are just prior to it.

Yes.
s/past/at/

**Comment by hgreving2304** [comment]
> > we need to be careful in the future that any thread has not been added to the global all_threads list as long as its tls magic is invalid.
> 
> I'm not sure I follow this: it sounds like you're adding an invariant that a thread on the all_threads list cannot have an invalid magic field, which is violated by design today. Do you mean something else? Maybe this needs to be re-worded.

Yes that's what it means. In the context of this change, if we did add a thread to the all_threads list that had its tls not yet initialized, we would potentially consider it a cloning thread in the signal handler. I think you're saying this would violate the design anyway? So I can just remove this comment.

**Comment by hgreving2304** [comment]
> > synchronize with an exiting threads
> 
> Incorrect grammar

Thanks, will fix.

**Comment by derekbruening** [comment]
> > > we need to be careful in the future that any thread has not been added to the global all_threads list as long as its tls magic is invalid.
> > 
> > I'm not sure I follow this: it sounds like you're adding an invariant that a thread on the all_threads list cannot have an invalid magic field, which is violated by design today. Do you mean something else? Maybe this needs to be re-worded.
> 
> Yes that's what it means. In the context of this change, if we did add a thread to the all_threads list that had its tls not yet initialized, we would potentially consider it a cloning thread in the signal handler. I think you're saying this would violate the design anyway? So I can just remove this comment.

No, I am saying the opposite: a thread that is on the all_threads list can certainly have an invalid magic field.  It happens all the time: for a brief window when a thread creates a child thread; when a thread goes native via various API's; etc.

**Comment by hgreving2304** [comment]
> > > > we need to be careful in the future that any thread has not been added to the global all_threads list as long as its tls magic is invalid.
> > > 
> > > 
> > > I'm not sure I follow this: it sounds like you're adding an invariant that a thread on the all_threads list cannot have an invalid magic field, which is violated by design today. Do you mean something else? Maybe this needs to be re-worded.
> > 
> > 
> > Yes that's what it means. In the context of this change, if we did add a thread to the all_threads list that had its tls not yet initialized, we would potentially consider it a cloning thread in the signal handler. I think you're saying this would violate the design anyway? So I can just remove this comment.
> 
> No, I am saying the opposite: a thread that is on the all_threads list can certainly have an invalid magic field. It happens all the time: for a brief window when a thread creates a child thread; when a thread goes native via various API's; etc.

* when a thread goes native via various API's; etc.: is this via SUSPEND and covered by the current case of i#2089 already? I think it is.
* for a brief window when a thread creates a child thread: from looking at "init", it looked like we're doing os_tls_init before add_thread.

File: core/unix/signal.c [hunk 1]
```diff
@@ -1276,7 +1290,8 @@ signal_thread_exit(dcontext_t *dcontext, bool other_thread)
         }
         info->num_pending = 0;
     }
-    if (!other_thread)
+    /* If no detach flag is set, we assume that this thread is on its way to exit. */
```
Comment (derekbruening) [inline core/unix/signal.c]: Please explain the reason (in the comment) for not restoring the mask: if I were reading this code with the existing comment I would assume it was just a performance optimization to skip the mask set and that I could remove this doing_detach check and it would still be correct.

Comment (derekbruening) [inline core/unix/signal.c]: The start/stop API used to support exiting a single thread: that is no longer the case, so I think this assumption is correct, and all the temporarily-native stuff does not come here.  Is there any way we can check it via an assert or sthg?  Probably not b/c it would have to be in app code.

File: core/unix/signal.c [hunk 2]
```diff
@@ -4838,12 +4853,24 @@ master_signal_handler_C(byte *xsp)
...
+    if (dcontext == NULL &&
+        /* Check for a temporarily-native thread we're synch-ing with. */
+        (sig == SUSPEND_SIGNAL
+#ifdef X86
+         || (INTERNAL_OPTION(safe_read_tls_init) &&
+             /* Check for whether this is a thread that makes a clone, in which case its
+              * magic field is temporarily invalid. It is also possible that it is a new
+              * thread on its way to init, but in this case the thread will not yet be
+              * known, because the thread will be added to the all_threads list after its
+              * tls has been initialized (i#2921).
```
Comment (derekbruening) [inline core/unix/signal.c]: I don't think this comment captures what we're doing because it doesn't talk about lock safety or native cases beyond clone.  We're trying to handle a signal in a temporarily-native thread, but it is unsafe to call thread_lookup from an arbitrary signal due to the lock it acquires.  If the magic field has our invalid sentinel we assume it's safe b/c it means the thread is either in app code (for native via API like DR_EMIT_GO_NATIVE) or in the clone window (where we know it doesn't grab a problematic lock from inspection of our own code).  Please generalize the comment: this is not just about a clone or init, and it should explain the lock problem.

File: core/unix/signal.c [hunk 3]
```diff
@@ -4838,12 +4853,24 @@ master_signal_handler_C(byte *xsp)
...
-            /* Check for a temporarily-native thread we're synch-ing with. */
-            tr = thread_lookup(get_sys_thread_id());
-            if (tr != NULL)
-                dcontext = tr->dcontext;
-        }
+    }
+    if (dcontext == NULL &&
+        /* Check for a temporarily-native thread we're synch-ing with. */
+        (sig == SUSPEND_SIGNAL
+#ifdef X86
```
Comment (derekbruening) [inline core/unix/signal.c]: I don't remember what ARM does (when will dcontext be NULL there): maybe worth a XXX comment.

**Comment by derekbruening** [comment]
> > when a thread goes native via various API's: etc.
>
> Is this via SUSPEND and covered by the current case of i#2089 already? I think it is.

No, for DR "API" means a client-exposed public interface, not an internal implementation detail like suspend signals.  This is client interfaces like DR_EMIT_GO_NATIVE.

> > for a brief window when a thread creates a child thread
>
> From looking at "init", it looked like we're doing os_tls_init before add_thread.

But the parent has a window where it is in all_threads but has an invalid magic.  I suggest rewording your invariant to make it clearer: I'm assuming you did not mean how I read it (because that does not hold) but I'm not sure what you do mean.

File: core/unix/os.c [hunk 4]
```diff
@@ -1323,6 +1323,14 @@ os_terminate_with_code(dcontext_t *dcontext, terminate_flags_t flags, int exit_c
     if (TEST(TERMINATE_CLEANUP, flags)) {
         /* we enter from several different places, so rewind until top-level kstat */
         KSTOP_REWIND_UNTIL(thread_measured);
+
+        /* This thread is on its way to exit, we are blocking all signals since any
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/unix/os.c]: Fixed.

File: core/unix/os_private.h [hunk 5]
```diff
@@ -271,6 +271,8 @@ void
 signal_thread_init(dcontext_t *dcontext, void *os_data);
 void
 signal_thread_exit(dcontext_t *dcontext, bool other_thread);
+void
+block_all_signals_except(kernel_sigset_t *oset, int num, ...);
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/unix/os_private.h]: Fixed.

File: core/unix/os.c [hunk 6]
```diff
@@ -6375,6 +6399,14 @@ handle_exit(dcontext_t *dcontext)
     }
     KSTOP(num_exits_dir_syscall);
 
+    /* This thread is on its way to exit, we are blocking all signals since any
+     * signal that reaches us now can be delayed until after the exit is complete.
+     * We may still receive a suspend signal for synchronization that we may need
+     * to reply to. If we are doing a detach, the thread's sigmask will be swapped
+     * back to the app's mask in signal_thread_exit (i#2921).
+     */
+    block_all_signals_except(NULL, 1, SUSPEND_SIGNAL);
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/unix/os.c]: Done

[the remark this answers is not in the document]
Comment (derekbruening) [reply core/unix/os.c]: Could this go into one of the *_thread_exit() routines?

[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/unix/os.c]: It could, only problem, we need to know whether it was terminate_via_kill, i.e. if terminating syscall is SYS_kill, so we had to pass this into the thread_exit function(s). This is why I didn't do it there.

File: core/unix/signal.c (hunk 1 above)
Comment (hgreving2304) [reply core/unix/signal.c]: Done

File: core/unix/signal.c (hunk 3 above)
Comment (hgreving2304) [reply core/unix/signal.c]: Do you want a new issue or even leave i#2921 open for this, or just XXX?

File: core/unix/signal.c (hunk 1 above)
Comment (hgreving2304) [reply core/unix/signal.c]: If DR_EMIT_NATIVE, then I guess no way to check? What about temporarily native e.g. for rseq? In any case those paths shouldn't come here indeed? Since this is called from cleanup_and_terminate?

File: core/unix/signal.c (hunk 2 above)
Comment (hgreving2304) [reply core/unix/signal.c]: Until a few minutes ago, I didn't know DR_EMIT_GO_NATIVE exists. I assume a temporarily-native thread has its tls magic field still set to valid?

Comment (hgreving2304) [reply core/unix/signal.c]: Comment refined.

File: core/unix/signal.c (hunk 3 above)
Comment (derekbruening) [reply core/unix/signal.c]: No strong opinion, whatever you want to do.

Comment (hgreving2304) [reply core/unix/signal.c]: Done.

File: core/unix/signal.c (hunk 1 above)
Comment (hgreving2304) [reply core/unix/signal.c]: (Resolving, re-open if necessary)

File: core/unix/os.c [hunk 7]
```diff
@@ -1301,6 +1301,28 @@ os_slow_exit(void)
     IF_NO_MEMQUERY(memcache_exit());
 }
 
+/* Helper function that calls cleanup_and_terminate after blocking most signals
+ *(i#2921).
+ */
+void
+cleanup_and_terminate_helper(dcontext_t *dcontext, int sysnum, ptr_uint_t sys_arg1,
```
Comment (derekbruening) [inline core/unix/os.c]: Please pick a more descriptive name: something like "block_cleanup_and_terminate" or "cleanup_and_terminate_blocked" or sthg which is more readable than "helper" which does not add anything to aid in understanding the code.  This is exported from this file which makes the name even more important.

File: core/unix/os.c [hunk 8]
```diff
@@ -6374,11 +6396,10 @@ handle_exit(dcontext_t *dcontext)
             exit_process ? "process" : "thread");
     }
     KSTOP(num_exits_dir_syscall);
-
```
Comment (derekbruening) [inline core/unix/os.c]: nit: No reason to remove the blank line which was setting this off from the block above, and is symmetric with the blank before that block

File: core/unix/os_private.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/unix/os_private.h]: I don't see any change?

File: core/unix/signal.c [hunk 9]
```diff
@@ -4838,12 +4860,28 @@ master_signal_handler_C(byte *xsp)
...
+         || (INTERNAL_OPTION(safe_read_tls_init) &&
+             /* Check for whether this is a thread with its invalid sentinel magic set.
+              * In this case, we assume that it is either a thread that is currently
+              * temporarily-native via API like DR_EMIT_GO_NATIVE, or a thread in the
+              * clone window. If the thread currently makes a clone, we will find it by
+              * looking it up. If the thread was just cloned, we will not find it yet.
+              * In both cases, we know by inspection of our own code, that it is safe to
```
Comment (derekbruening) [inline core/unix/signal.c]: Kind of confusing as to what "both" refers to: "either ... temp-native ... or ... clone window", or "makes a clone...was just cloned".  I'm not sure what the two prior sentences "If...we will find it..." and "If...we will not find it" are adding: I would just remove them.

File: core/unix/os_private.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/unix/os_private.h]: Declaration was missing, fixed

File: core/unix/os.c (hunk 7 above)
Comment (hgreving2304) [reply core/unix/os.c]: Done.

File: core/unix/os.c (hunk 8 above)
Comment (hgreving2304) [reply core/unix/os.c]: Fixed.

File: core/unix/signal.c (hunk 9 above)
Comment (hgreving2304) [reply core/unix/signal.c]: Done
